service/history/callback_queue_active_task_executor.go

```diff
@@ -0,0 +1,293 @@
+	case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_CANCELED:
+		return &nexus.OperationCompletionUnsuccessful{
+			State:   nexus.OperationStateCanceled,
+			Failure: &nexus.Failure{Message: "operation canceled"},
+		}, nil
+	case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_TERMINATED:
+		return &nexus.OperationCompletionUnsuccessful{
+			State:   nexus.OperationStateFailed,
+			Failure: &nexus.Failure{Message: "operation terminated"},
+		}, nil
+	case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_TIMED_OUT:
+		return &nexus.OperationCompletionUnsuccessful{
+			State:   nexus.OperationStateFailed,
+			Failure: &nexus.Failure{Message: "operation exceeded internal timeout"},
+		}, nil
+	}
+	return nil, serviceerror.NewUnimplemented(fmt.Sprintf("invalid workflow execution status: %v", ce.GetEventType()))
+}
+
+func (t *callbackQueueActiveTaskExecutor) processNexusCallbackTask(ctx context.Context, task *tasks.CallbackTask, url string, completion nexus.OperationCompletion) error {
+	request, err := nexus.NewCompletionHTTPRequest(ctx, url, completion)
+	if err != nil {
+		return serviceerror.NewInternal(fmt.Sprintf("failed to construct Nexus request: %v", err))
+	}
+
+	caller := t.callerGetter(task.NamespaceID, task.DestinationAddress)
+	response, err := caller(request)
+	return t.accessCallback(ctx, task, func(wfCtx workflow.Context, env workflow.MutableState, callback *persistencespb.CallbackInfo) error {
+		err = func() error {
+			if err != nil {
+				return callbacks.TransitionAttemptFailed.Apply(callback, callbacks.EventAttemptFailed(err), env)
+			}
+			if response.StatusCode >= 200 && response.StatusCode < 300 {
+				return callbacks.TransitionSucceeded.Apply(callback, callbacks.EventSucceeded{}, env)
+			}
+			if response.StatusCode >= 400 && response.StatusCode < 500 && !slices.Contains(retryable4xxErrorTypes, response.StatusCode) {
+				return callbacks.TransitionFailed.Apply(callback, callbacks.EventFailed(err), env)
+			}
+			return callbacks.TransitionAttemptFailed.Apply(callback, callbacks.EventAttemptFailed(fmt.Errorf("request failed with: %v", response.Status)), env)
+		}()
+
+		return wfCtx.UpdateWorkflowExecutionAsActive(ctx, t.shard)
+	})
+}
+
+// accessCallback loads mutable state, and extracts the callback for the given task.
+// accessFn is called while holding a lock on the workflow context.
+func (t *callbackQueueActiveTaskExecutor) accessCallback(
+	ctx context.Context,
+	task *tasks.CallbackTask,
+	accessFn func(workflow.Context, workflow.MutableState, *persistencespb.CallbackInfo) error,
+) (retErr error) {
+	wfCtx, release, err := getWorkflowExecutionContextForTask(ctx, t.shard, t.workflowCache, task)
+	if err != nil {
+		return err
+	}
+	accessed := false
+	defer func() {
+		if !accessed {
+			// Before we run the accessFn, all errors are fine to ignore and it is safe to release with no error to
+			// prevent mutable state from being unloaded from the cache.
+			release(nil)
+		}
+		release(retErr)
+	}()
+
+	for attempt := 1; ; attempt++ {
```

Comment:
do I understand correctly that this loop will do 2 iterations at most?

Reply:
Yes, I'll add a comment to clarify.